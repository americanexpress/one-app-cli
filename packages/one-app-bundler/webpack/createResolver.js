/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const {
  NodeJsInputFileSystem,
  CachedInputFileSystem,
  ResolverFactory,
} = require('enhanced-resolve');

// This enhanced resolver ensures that the code for a lib that is exposed is the
// same code that is used within the bundle. This bug would happen if an exposed
// lib had a module or browser field in addition to the main field in its
// package.json.

module.exports = function createResolver({ mainFields, resolveToContext = false }) {
  const enhancedResolver = ResolverFactory.createResolver({
    fileSystem: new CachedInputFileSystem(new NodeJsInputFileSystem(), 4000),
    useSyncFileSystemCalls: true,
    extensions: ['.js', '.jsx', '.json'],
    mainFields,
    resolveToContext,
  });

  const trailingSlash = process.platform === 'win32' ? '\\' : '/';
  // if resolveToContext add a trailing slash to indicate the value is a folder rather than  a file
  return (request) => `${enhancedResolver.resolveSync({}, __dirname, request)}${resolveToContext ? trailingSlash : ''}`;
};
