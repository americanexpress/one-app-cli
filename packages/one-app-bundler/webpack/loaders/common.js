/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const path = require('path');
const dartSass = require('sass');
const Fiber = require('fibers');
const getConfigOptions = require('../../utils/getConfigOptions');

const packageRoot = process.cwd();

const cssLoader = ({ name = '', importLoaders = 2 } = {}) => ({
  loader: 'css-loader',
  options: {
    importLoaders,
    modules: {
      localIdentName: `${name && `${name}__`}[name]__[local]___[hash:base64:5]`,
      // getLocalIdent is a function that allows you to specify a function to generate the classname
      // The documentation can be found here:
      // https://github.com/webpack-contrib/css-loader#getlocalident

      // The below function returns the classnames as is if the resourcePath includes node_modules
      // if it doesn't it returns null allowing localIdentName to define the classname
      // eslint-disable-next-line max-params, no-unused-vars
      getLocalIdent: (loaderContext, localIdentName, localName, options) => (
        loaderContext.resourcePath.includes('node_modules') ? localName : null
      ),
    },
  },
});

const purgeCssLoader = () => {
  const configOptions = getConfigOptions();
  const whitelistPatterns = configOptions.purgecss.whitelistPatterns
    ? configOptions.purgecss.whitelistPatterns.map((pattern) => new RegExp(pattern, 'i'))
    : [];
  const whitelistPatternsChildren = configOptions.purgecss.whitelistPatternsChildren
    ? configOptions.purgecss.whitelistPatternsChildren.map((pattern) => new RegExp(pattern, 'i'))
    : [/:global$/];
  if (configOptions.purgecss.disabled) return [];
  return [{
    loader: '@americanexpress/purgecss-loader',
    options: {
      paths: [path.join(packageRoot, 'src/**/*.{js,jsx}'), ...configOptions.purgecss.paths || []],
      extractors: configOptions.purgecss.extractors || [],
      fontFace: configOptions.purgecss.fontFace || false,
      keyframes: configOptions.purgecss.keyframes || false,
      variables: configOptions.purgecss.variables || false,
      whitelist: configOptions.purgecss.whitelist || [],
      whitelistPatterns,
      whitelistPatternsChildren,
    },
  }];
};

const sassLoader = () => ({
  loader: 'sass-loader',
  options: {
    implementation: dartSass,
    sassOptions: {
      fiber: Fiber,
    },
  },
});

const babelLoader = (babelEnv) => ({
  loader: 'babel-loader',
  options: {
    extends: path.join(packageRoot, '.babelrc'),
    envName: babelEnv,
    cacheDirectory: path.join(packageRoot, '.build-cache'),
  },
});

module.exports = {
  babelLoader,
  cssLoader,
  purgeCssLoader,
  sassLoader,
};
