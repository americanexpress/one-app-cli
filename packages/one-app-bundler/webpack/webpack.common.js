/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/**
 * Webpack Base Configuration
 */

const webpack = require('webpack');
const TerserPlugin = require('terser-webpack-plugin');
const ESLintPlugin = require('eslint-webpack-plugin');
const validateNodeEnv = require('../utils/validateNodeEnv');

validateNodeEnv();

const nodeEnvIsProduction = process.env.NODE_ENV === 'production';

const prodPlugins = [
  new webpack.LoaderOptionsPlugin({
    minimize: true,
    debug: false,
  }),
];

const plugins = [
  new webpack.EnvironmentPlugin([
    'NODE_ENV',
  ]),
  new ESLintPlugin({
    allowInlineConfig: false,
    ignore: false,
    failOnError: true,
    useEslintrc: false,
    overrideConfig: {
      parserOptions: {
        ecmaVersion: 6,
        sourceType: 'module',
        ecmaFeatures: { modules: true },
      },
      rules: {
        'no-restricted-modules': ['error', 'create-react-class'],
        'no-restricted-imports': ['error', 'create-react-class'],
        '@americanexpress/one-app/no-app-config-on-client': 'error',
      },
      plugins: ['@americanexpress/one-app'],
    },
  }),
  ...(nodeEnvIsProduction ? prodPlugins : []),
];

module.exports = {
  profile: true,
  devtool: nodeEnvIsProduction ? false : 'source-map',
  optimization: {
    minimize: nodeEnvIsProduction,
    minimizer: [
      new TerserPlugin({
        test: /\.jsx?$/i,
        extractComments: false,
        terserOptions: {
          compress: {
            drop_console: true,
          },
          keep_fnames: true,
        },
      }),
    ],
  },
  mode: nodeEnvIsProduction ? 'production' : 'development',
  module: {
    rules: [
      {
        test: /\.(ttf|eot|svg|png|jpg)(\?.*)?$/,
        use: [{ loader: 'url-loader' }],
      },
      {
        test: /\.(woff|woff2)(\?.*)?$/,
        use: [{
          loader: 'url-loader',
          options: {
            limit: 10000,
            mimetype: 'application/font-woff',
          },
        }],
      },
    ],
  },
  externals: {
    '@americanexpress/one-app-router': {
      var: 'OneAppRouter',
      commonjs2: '@americanexpress/one-app-router',
    },
    'create-shared-react-context': {
      var: 'CreateSharedReactContext',
      commonjs2: 'create-shared-react-context',
    },
    holocron: {
      var: 'Holocron',
      commonjs2: 'holocron',
    },
    react: {
      var: 'React',
      commonjs2: 'react',
    },
    'react-dom': {
      var: 'ReactDOM',
      commonjs2: 'react-dom',
    },
    redux: {
      var: 'Redux',
      commonjs2: 'redux',
    },
    'react-redux': {
      var: 'ReactRedux',
      commonjs2: 'react-redux',
    },
    reselect: {
      var: 'Reselect',
      commonjs2: 'reselect',
    },
    immutable: {
      var: 'Immutable',
      commonjs2: 'immutable',
    },
    '@americanexpress/one-app-ducks': {
      var: 'OneAppDucks',
      commonjs2: '@americanexpress/one-app-ducks',
    },
    'holocron-module-route': {
      var: 'HolocronModuleRoute',
      commonjs2: 'holocron-module-route',
    },
    'prop-types': {
      var: 'PropTypes',
      commonjs2: 'prop-types',
    },
    'react-helmet': {
      var: 'ReactHelmet',
      commonjs2: 'react-helmet',
    },
  },
  plugins,
};
