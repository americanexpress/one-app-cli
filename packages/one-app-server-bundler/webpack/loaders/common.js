/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const path = require('node:path');
const dartSass = require('sass');

const packageRoot = process.cwd();

const cssLoader = ({ name = '', importLoaders = 2 } = {}) => ({
  loader: 'css-loader',
  options: {
    importLoaders,
    modules: {
      localIdentName: `${name && `${name}__`}[name]__[local]___[hash:base64:5]`,
      // getLocalIdent is a function that allows you to specify a function to generate the classname
      // The documentation can be found here:
      // https://github.com/webpack-contrib/css-loader#getlocalident

      // The below function returns the classnames as is if the resourcePath includes node_modules
      // if it doesn't it returns null allowing localIdentName to define the classname
      getLocalIdent: (loaderContext, localIdentName, localName) => (
        loaderContext.resourcePath.includes('node_modules') ? localName : null
      ),
    },
  },
});

const sassLoader = () => ({
  loader: 'sass-loader',
  options: {
    implementation: dartSass,
  },
});

const babelLoader = (babelEnv) => ({
  loader: 'babel-loader',
  options: {
    extends: path.join(packageRoot, '.babelrc'),
    envName: babelEnv,
    cacheDirectory: path.join(packageRoot, '.build-cache'),
  },
});

module.exports = {
  babelLoader,
  cssLoader,
  sassLoader,
};
