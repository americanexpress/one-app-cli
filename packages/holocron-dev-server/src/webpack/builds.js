/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import webpack from 'webpack';

import { createHolocronModuleWebpackConfig, createExternalsDllWebpackConfig } from './configs/bundles';
import { modulesLibraryVarName } from './helpers';
import { ufs } from '../utils';
import {
  logError,
  logWebpackStatsWhenDone,
  logWhenWebpackInvalid,
  logExternalsBuilding,
} from '../utils/logs';

export function buildWebpackConfig(config) {
  return new Promise((resolve, reject) => {
    webpack(config).run((err, stats) => {
      if (err) {
        logError(err);
        reject(err);
      } else {
        resolve(stats);
      }
    });
  });
}

export function buildModuleExternalsDllBundle(config = {}) {
  const { externals: externalsToPackageDll = [] } = config;

  if (!(externalsToPackageDll.length > 0)) return Promise.resolve();

  return buildWebpackConfig(
    createExternalsDllWebpackConfig({
      entries: externalsToPackageDll,
    })
  ).then((stats) => {
    logExternalsBuilding(externalsToPackageDll);
    return stats;
  });
}

export function createHotHolocronCompiler({
  modules,
  externals,
  environmentVariables,
  webpackConfigPath,
}) {
  const holocronWebpackConfig = createHolocronModuleWebpackConfig({
    hot: true,
    modules,
    externals,
    environmentVariables,
    webpackConfigPath,
  });
  const compiler = webpack(holocronWebpackConfig);
  compiler.inputFileSystem = ufs;
  compiler.resolvers.normal.fileSystem = compiler.inputFileSystem;
  compiler.resolvers.context.fileSystem = compiler.inputFileSystem;
  compiler.hooks.done.tap(modulesLibraryVarName, logWebpackStatsWhenDone);
  compiler.hooks.invalid.tap(modulesLibraryVarName, logWhenWebpackInvalid);
  return compiler;
}
