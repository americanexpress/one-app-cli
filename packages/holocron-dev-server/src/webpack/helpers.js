/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { version } from 'webpack';

export const modulesLibraryVarName = '__holocron_modules__';
export const externalsLibraryVarName = '__externals__';

export const modulesFilename = '[name]/[name].js';
export const assetModuleFilename = 'assets/[name].[ext]';

export const jsxTest = /\.jsx?$/i;
export const fileTest = /\.(woff|woff2|ttf|eot|svg|png|jpg|jpeg|gif|webm)(\?.*)?$/;
export const cssTest = /\.(sa|sc|c)ss$/;
export const nodeModulesPattern = /node_modules/;

export function getWebpackVersion() {
  return parseInt(version, 10);
}

export function createExternalEntry([packageName, varName]) {
  return {
    [packageName]: {
      commonjs2: packageName,
      ...varName
        ? {
          var: varName,
          root: varName,
        }
        : {},
    },
  };
}

export function createOneAppExternals(additionalExternals = []) {
  // TODO: pull this from one-app-bundler when merging
  return [
    ['@americanexpress/one-app-ducks', 'OneAppDucks'],
    ['@americanexpress/one-app-router', 'OneAppRouter'],
    ['create-shared-react-context', 'CreateSharedReactContext'],
    ['holocron', 'Holocron'],
    ['holocron-module-route', 'HolocronModuleRoute'],
    ['immutable', 'Immutable'],
    ['prop-types', 'PropTypes'],
    ['react', 'React'],
    ['react-dom', 'ReactDOM'],
    ['react-helmet', 'ReactHelmet'],
    ['react-redux', 'ReactRedux'],
    ['redux', 'Redux'],
    ['reselect', 'Reselect'],
    ...additionalExternals.map((external) => (Array.isArray(external) ? external : [external])),
  ]
    .map(createExternalEntry)
    .reduce((map, next) => ({ ...map, ...next }), {});
}

const createHotModuleEntry = ({ moduleName, modulePath }, hot) => ({
  [moduleName]: [
    ...hot
      ? [require.resolve('webpack-hot-middleware/client'), require.resolve('react-refresh/runtime')]
      : [],
    `${modulePath}/src/index.js`,
  ],
});

export function createHolocronModuleEntries({ modules = [], hot = false } = {}) {
  return modules
    .map((module) => createHotModuleEntry(module, hot))
    .reduce((map, next) => ({ ...map, ...next }), {});
}
