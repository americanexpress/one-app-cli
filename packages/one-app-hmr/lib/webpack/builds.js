/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import webpack from 'webpack';

import { createHolocronModuleWebpackConfig, createExternalsDllWebpackConfig } from './configs';
import { modulesLibraryVarName } from './helpers';
import { ufs } from '../utils';
import {
  error,
  logWebpackStatsWhenDone,
  logWhenWebpackInvalid,
  logExternalsBuilding,
} from '../utils/logs';

export function runWebpackCompiler(compiler) {
  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) {
        error(err);
        reject(err);
      } else {
        resolve(stats);
      }
    });
  });
}

export function buildWebpackConfig(config) {
  return runWebpackCompiler(webpack(config));
}

export function buildModuleExternalsDllBundle(config = {}) {
  const { externals: externalsToPackageDll = [], name } = config;

  if (!(externalsToPackageDll.length > 0)) return Promise.resolve();

  return buildWebpackConfig(
    createExternalsDllWebpackConfig({
      name,
      entries: externalsToPackageDll,
    })
  ).then((stats) => {
    logExternalsBuilding(externalsToPackageDll);
    return stats;
  });
}

export function createHotHolocronCompiler({
  modules,
  externals,
  environmentVariables,
  babelConfig,
  webpackConfigPath,
}) {
  const holocronWebpackConfig = createHolocronModuleWebpackConfig({
    hot: true,
    modules,
    externals,
    environmentVariables,
    babelConfig,
    webpackConfigPath,
  });
  const compiler = webpack(holocronWebpackConfig);
  compiler.inputFileSystem = ufs;
  compiler.resolvers.normal.fileSystem = compiler.inputFileSystem;
  compiler.resolvers.context.fileSystem = compiler.inputFileSystem;
  compiler.hooks.done.tap(modulesLibraryVarName, logWebpackStatsWhenDone);
  compiler.hooks.invalid.tap(modulesLibraryVarName, logWhenWebpackInvalid);
  return compiler;
}
