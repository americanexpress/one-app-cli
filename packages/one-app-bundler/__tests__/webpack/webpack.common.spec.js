/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

jest.mock('../../utils/validateNodeEnv');

describe('webpack/one-amex.base', () => {
  let nodeEnv;

  beforeAll(() => {
    nodeEnv = process.env.NODE_ENV;
  });

  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
  });

  afterEach(() => {
    process.env.NODE_ENV = nodeEnv;
  });

  it('should validate the NODE_ENV', () => {
    const validateNodeEnv = require('../../utils/validateNodeEnv');
    require('../../webpack/webpack.common');
    expect(validateNodeEnv).toHaveBeenCalled();
  });

  it('should add some plugins and set the profile and minimize flags to true', () => {
    process.env.NODE_ENV = 'production';
    const webpackConfig = require('../../webpack/webpack.common');
    expect(webpackConfig).toEqual({
      devtool: false,
      profile: true,
      externals: expect.any(Object),
      module: expect.any(Object),
      mode: 'production',
      plugins: expect.any(Array),
      optimization: {
        minimize: true,
        minimizer: expect.any(Array),
      },
    });
  });

  it('should enable source maps in development', () => {
    process.env.NODE_ENV = 'development';
    const webpackConfig = require('../../webpack/webpack.common');
    expect(webpackConfig).toEqual({
      devtool: 'source-map',
      profile: true,
      externals: expect.any(Object),
      module: expect.any(Object),
      mode: 'development',
      plugins: expect.any(Array),
      optimization: {
        minimize: false,
        minimizer: expect.any(Array),
      },
    });
  });

  it('should only include the EnvironmentPlugin in development', () => {
    process.env.NODE_ENV = 'development';
    const webpackConfig = require('../../webpack/webpack.common');
    expect(webpackConfig.plugins).toHaveLength(2);
    expect(webpackConfig.plugins).toMatchSnapshot();
  });

  it('should add loader options plugin in production', () => {
    process.env.NODE_ENV = 'production';
    const webpackConfig = require('../../webpack/webpack.common');
    expect(webpackConfig.plugins).toHaveLength(3);
    expect(webpackConfig.plugins).toMatchSnapshot();
  });
});
