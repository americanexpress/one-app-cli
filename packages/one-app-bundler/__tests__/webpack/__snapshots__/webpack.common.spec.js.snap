// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack/one-amex.base should add loader options plugin in production 1`] = `
Array [
  EnvironmentPlugin {
    "defaultValues": Object {},
    "keys": Array [
      "NODE_ENV",
    ],
  },
  LoaderOptionsPlugin {
    "options": Object {
      "debug": false,
      "minimize": true,
      "test": Object {
        "test": [Function],
      },
    },
  },
]
`;

exports[`webpack/one-amex.base should only include the EnvironmentPlugin in development 1`] = `
Array [
  EnvironmentPlugin {
    "defaultValues": Object {},
    "keys": Array [
      "NODE_ENV",
    ],
  },
]
`;
